examples/example_compiler_gym_service/service_py/example_service.py

```diff
@@ -0,0 +1,225 @@
+# dynamic action space.
+ACTION_SPACE = proto.ActionSpace(
+    name="default",
+    action=[
+        "a",
+        "b",
+        "c",
+    ],
+)
+
+# A list of observation spaces supported by this service. Each of these
+# ObservationSpace protos describes an observation space.
+OBSERVATION_SPACES = [
+    proto.ObservationSpace(
+        name="ir",
+        string_size_range=proto.ScalarRange(min=proto.ScalarLimit(value=0)),
+        deterministic=True,
+        platform_dependent=False,
+        default_value=proto.Observation(string_value=""),
+    ),
+    proto.ObservationSpace(
+        name="features",
+        int64_range_list=proto.ScalarRangeList(
+            range=[
+                proto.ScalarRange(
+                    min=proto.ScalarLimit(value=-100), max=proto.ScalarLimit(value=100)
+                ),
+                proto.ScalarRange(
+                    min=proto.ScalarLimit(value=-100), max=proto.ScalarLimit(value=100)
+                ),
+                proto.ScalarRange(
+                    min=proto.ScalarLimit(value=-100), max=proto.ScalarLimit(value=100)
+                ),
+            ]
+        ),
+    ),
+    proto.ObservationSpace(
+        name="runtime",
+        scalar_double_range=proto.ScalarRange(min=proto.ScalarLimit(value=0)),
+        deterministic=False,
+        platform_dependent=True,
+        default_value=proto.Observation(
+            scalar_double=0,
+        ),
+    ),
+]
+
+
+class CompilationSession(object):
+    """Represents an instance of an interactive compilation session."""
+
+    def __init__(self, benchmark: str):
+        # Do any of the set up required to start a compilation "session".
+        self.benchmark = benchmark
+
+    def step(self, request: proto.StepRequest, context) -> proto.StepReply:
+        reply = proto.StepReply()
+
+        # Apply a list of actions from the user. Each value is an index into the
+        # ACTIONS_SPACE.action list.
+        for action in request.action:
+            logging.debug("Apply action %d", action)
+            if action < 0 or action >= len(ACTION_SPACE.action):
+                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+                context.set_details("Out-of-range")
+                return
+
+        # Compute a list of observations from the user. Each value is an index
+        # into the OBSERVATION_SPACES list.
+        for observation_space in request.observation_space:
+            logging.debug("Compute observation %d", observation_space)
+            observation = reply.observation.add()
+            if observation_space == 0:  # ir
+                observation.string_value = "Hello, world!"
+            elif observation_space == 1:  # features
+                observation.int64_list.value[:] = [0, 0, 0]
+            elif observation_space == 1:  # runtime
+                observation.scalar_double = 0
+
+        return reply
+
+
+class ExampleCompilerGymService(proto.CompilerGymServiceServicer):
```

Comment:
this could probably be hidden away and take in a class that implements `CompilationSession` (with the addition of class methods to get `ACTION_SPACE` and `OBSERVATION_SPACES`)

Reply:
I agree, this would be really nice. I think once we start adding support for multiple compilers we should refactor away the boilerplate for session management / handling RPC requests